### PR TITLE
Split service discovery into layered components

### PR DESCRIPTION
**Pull request by olix0r**

### CODE REVIEW

File: src/svc.rs [hunk 1]
```diff
@@ -103,6 +104,14 @@ impl<S> Stack<S> {
         self.push(TimeoutLayer::new(timeout))
     }
 
+    /// Validates that this stack serves T-typed targets.
+    pub fn makes<T>(self) -> Self
```
Comment (adleong) [inline src/svc.rs]: unused in this PR?

File: src/proxy/http/profiles/router.rs [hunk 2]
```diff
@@ -133,24 +133,23 @@ where
     }
 }
 
-impl<G, Inner, RouteLayer, RouteBody, InnerBody, Target, InnerSvc, RouteMake, RouteSvc>
-    svc::Service<Target> for MakeSvc<G, Inner, RouteLayer, RouteBody, InnerBody>
+impl<G, Inner, RouteLayer, RouteBody, InnerBody, Target, RouteSvc> svc::Service<Target>
```
Comment (adleong) [inline src/proxy/http/profiles/router.rs]: Nice cleanups in this file, but I don't see how it's related to the rest of the PR

File: src/svc.rs (hunk 1 above)
Comment (olix0r) [reply src/svc.rs]: yeah, but it's useful. this module is for useful things when working with stacks.

File: src/proxy/http/profiles/router.rs (hunk 2 above)
Comment (olix0r) [reply src/proxy/http/profiles/router.rs]: i needed to make some changes below and was fighting type signatures to make it all happy... i can probably back some of this out...

**Comment by olix0r** [comment]
Replaced by #341

**Pull request closed without merging**